### PR TITLE
Fixes 1594-hide-discussion and 1595-edit-discussion. Adds e2e tests

### PR DESCRIPTION
**Pull request by pepoospina**
This a single PR with a single squashed commit that combines the previous two PRs to edit and hide discussions.

### CODE REVIEW

**Comment by pepoospina** [comment]
try it with:

in the root folder
```npm run up-test``` 

Then in functions folder
```
npm run test-e2e
```

File: package.json [hunk 1]
```diff
@@ -0,0 +1,25 @@
+{
```
Comment (ariel-dao) [inline package.json:1]: remove file package json needs to stay in functions folder

File: functions/src/shared/schemas/updateDiscussion.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import { object, string, InferType } from "yup";
+
+// The validation schema for creating a discussion
+export const updateDiscussionDataValidationScheme = object({
+  discussionId: string().required(),
```
Comment (ariel-dao) [inline functions/src/shared/schemas/updateDiscussion.ts:5]: why is updating requires all ? change to optional and check validation inside

File: functions/src/shared/schemas/hideContent.ts [hunk 3]
```diff
@@ -3,7 +3,5 @@ import { MODERATION_TYPES } from '../enums/MODERATION_TYPES';
 
 export const hideContentDataValidationScheme = yup.object({
     itemId: yup.string().required(),
-    commonId: yup.string().required(),
```
Comment (ariel-dao) [inline functions/src/shared/schemas/hideContent.ts:6]: why?

File: functions/src/shared/schemas/showContent.ts [hunk 4]
```diff
@@ -3,8 +3,6 @@ import { MODERATION_TYPES } from '../enums/MODERATION_TYPES';
 
 export const showContentDataValidationScheme = yup.object({
     itemId: yup.string().required(),
-    commonId: yup.string().required(),
```
Comment (ariel-dao) [inline functions/src/shared/schemas/showContent.ts:6]: why?

File: functions/src/shared/interfaces/discussions/Discussion.ts [hunk 5]
```diff
@@ -30,6 +30,11 @@ export interface Discussion extends BaseEntity {
    */
   lastMessage: Timestamp;
 
+  /**
+   * When was the discussion edited
+   */
+  editedAt?: Timestamp
```
Comment (ariel-dao) [inline functions/src/shared/interfaces/discussions/Discussion.ts:36]: is this required? if not use updated at

File: functions/src/moderation/index.ts [hunk 6]
```diff
@@ -29,9 +29,8 @@ router.post('/hide', async (req: Request, res: Response, next: NextFunction) =>
   await responseExecutor(
     async () => {
       return await hideContent({
-        ...req.body,
-        userId: req.user.uid
-      });
+        ...req.body
+      }, req.user.uid);
```
Comment (ariel-dao) [inline functions/src/moderation/index.ts:33]: why change this function?

File: functions/src/moderation/business/showContent.ts [hunk 7]
```diff
@@ -30,11 +32,13 @@ const checkIfAlreadyVisibile = async (itemId: string, itemType: MODERATION_TYPES
   }
 }
 
-export const showContent = async (showContentPayload: ShowContentPayload) => {
+export const showContent = async (showContentPayload: ShowContentPayload, userId: string) => {
```
Comment (ariel-dao) [inline functions/src/moderation/business/showContent.ts:35]: why change this function?

File: functions/src/moderation/business/hideContent.ts [hunk 8]
```diff
@@ -28,11 +30,11 @@ const checkIfAlreadyHidden = async (itemId: string, itemType: MODERATION_TYPES)
   }
 }
 
-export const hideContent = async (hideContentPayload: HideContentPayload) => {
+export const hideContent = async (hideContentPayload: HideContentPayload, userId: string) => {
```
Comment (ariel-dao) [inline functions/src/moderation/business/hideContent.ts:33]: why change this function?

File: functions/src/discussion/services/discussion/update.ts [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+import { discussionDb } from "../../db";
+
+interface UpdateDiscussionService {
+  discussionId: string
+  title: string;
+  message: string;
+}
+
+export const updateDiscussionService = async (discussion: UpdateDiscussionService) => {
+  await discussionDb.updateDiscussion(discussion.discussionId, {
```
Comment (ariel-dao) [inline functions/src/discussion/services/discussion/update.ts:11]: update only changed no need to resend all data

File: functions/src/discussion/controllers/discussionMessage/create.ts [hunk 10]
```diff
@@ -4,6 +4,8 @@ import { createDiscussionMessageDataValidationScheme } from "../../../shared/sch
 import { createDiscussionMessageService } from "../../services";
 import { discussionDb } from "../../db";
 
+export const successMessage = 'Discussion Message created successfully';
```
Comment (ariel-dao) [inline functions/src/discussion/controllers/discussionMessage/create.ts:7]: I get the change. but this PR is not the place for it. Keep it but please refrain for adding unrelated stuff

File: functions/src/discussion/controllers/discussion/update.ts [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+import { discussionDb } from "../../db";
+import { updateDiscussionDataValidationScheme } from "../../../shared/schemas/updateDiscussion";
+import { updateDiscussionService } from "../../../discussion/services/discussion/update";
+
+export const successMessage = "Discussion updated successfully";
+
+export const updateDiscussion = async (req: Request, res: Response, next: NextFunction) =>
+  await responseExecutor(
+    async () => {
+      const validated = await updateDiscussionDataValidationScheme.validate(
```
Comment (ariel-dao) [inline functions/src/discussion/controllers/discussion/update.ts:13]: one line it

File: functions/src/discussion/controllers/discussion/update.ts [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+        true
+      );
+
+      if (discussion.ownerId !== req.user.uid)
+        throw new CommonError("Not discussion owner", {
+          ownerId: req.user.uid,
+          discussionId: validated.discussionId,
+        });
+
+      // circle visibility should remain the same
```
Comment (ariel-dao) [inline functions/src/discussion/controllers/discussion/update.ts:28]: no reason to comment this, enforce it with validation checks and types

File: env/env.sh [hunk 13]
```diff
@@ -1,6 +1,4 @@
 #!/bin/bash
-cd ..
```
Comment (ariel-dao) [inline env/env.sh:2]: remove put back in functions

File: functions/package.json [hunk 14]
```diff
@@ -55,11 +50,13 @@
     "@types/express": "^4.17.15",
     "@types/express-pino-logger": "^4.0.3",
     "@types/google-libphonenumber": "^7.4.23",
+    "@types/ioredis": "^5.0.0",
```
Comment (ariel-dao) [inline functions/package.json:53]: why?

File: functions/package.json [hunk 15]
```diff
@@ -9,16 +9,11 @@
     "node": "16"
   },
   "scripts": {
-    "dev": "NODE_ENV=local FIRESTORE_EMULATOR_HOST=localhost:8080 firebase emulators:start",
-    "test": "NODE_ENV=test mocha --check-leaks -r ts-node/register -r tsconfig-paths/register \"src/__tests__/**/*.spec.ts\"",
-    "build": "npm test && tsc -p tsconfig.prod.json",
-    "deploy:dev": "npm run use:dev && ../env/env.sh -check && firebase deploy --only functions",
-    "deploy:staging": "npm run use:staging && ../env/env.sh -check && firebase deploy --only functions",
-    "deploy:production": "npm run use:production && ../env/env.sh -check && firebase deploy --only functions",
```
Comment (ariel-dao) [inline functions/package.json:17]: why did you remove all deployemnt scripts?  and hooks? and use env?

File: functions/src/__e2e_tests__/commons.test.ts [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+import "./index.test";
+import { http } from "./http.connector";
+import { expect } from "chai";
+
+describe("common", () => {
+  const NAME = "My Common";
+
+  it("create", async () => {
+    const common = await createCommon(http, { name: NAME });
+    expect(common.name).to.be.eq(NAME);
```
Comment (ariel-dao) [inline functions/src/__e2e_tests__/commons.test.ts:15]: validated all the entity, and if you already call it e2e, make sure to actually make it e2e but calling the controller and emulating the function and firestore

File: functions/src/__e2e_tests__/discussion.test.ts [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+  editDiscussionMessage,
+  getDiscussion,
+  postDiscussionMessage,
+  readFullDiscussion,
+} from "./utils/discussion";
+import { hideItem, showItem } from "./utils/moderate";
+
+describe("discussions", () => {
+  const httpAlice = new HttpConnector(alice);
+  // const httpBob = new HttpConnector(bob);
```
Comment (ariel-dao) [inline functions/src/__e2e_tests__/discussion.test.ts:25]: remove

File: functions/src/__e2e_tests__/discussion.test.ts [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+
+  it("create", async () => {
+    expect(typeof commonId).to.be.eq("string");
+  });
+
+  describe("create", () => {
+    let discussionId: string;
+    let messageId: string;
+
+    const DISCUSSION = {
```
Comment (ariel-dao) [inline functions/src/__e2e_tests__/discussion.test.ts:42]: discussion should be of type Discussion, it'll make things consistent

File: functions/src/__e2e_tests__/discussion.test.ts [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+    before(async () => {
+      const create = await createDiscussion(
+        commonId,
+        DISCUSSION.title,
+        DISCUSSION.message,
+        httpAlice
+      );
+      discussionId = create.id;
+
+      expect(create.title).to.be.eq(DISCUSSION.title);
```
Comment (ariel-dao) [inline functions/src/__e2e_tests__/discussion.test.ts:63]: validated all the entity, and if you already call it e2e, make sure to actually make it e2e but calling the controller and emulating the function and firestore

File: functions/src/__e2e_tests__/discussion.test.ts [hunk 20]
```diff
@@ -0,0 +1,175 @@
...
+      expect(create.message).to.be.eq(DISCUSSION.message);
+      // expect(create.message).toBe(DISCUSSION_CREATED_SUCCESS);
+
+      const msg = await postDiscussionMessage(
```
Comment (ariel-dao) [inline functions/src/__e2e_tests__/discussion.test.ts:67]: different test

File: functions/src/__e2e_tests__/discussion.test.ts [hunk 21]
```diff
@@ -0,0 +1,175 @@
...
+        commonId,
+        httpAlice
+      );
+
+      messageId = msg.id;
+
+      expect(msg.message).to.be.eq(MESSAGE_CREATED_SUCCESS);
+    });
+
+    it("created correctly", async () => {
```
Comment (ariel-dao) [inline functions/src/__e2e_tests__/discussion.test.ts:80]: we should really use more actionable and contextual names, each test should change one thing. I suggest to write in AAA. arrange, act, assert

File: package.json (hunk 1 above)
Comment (pepoospina) [reply package.json:1]: Functions are executable code, I'd keep their `package.json` limited to whatever is needed to execute that code, independently of the deployment process.

The deployment process involves the sdk, firestore and different environments configuration. It's no coincidence that all deployment related scripts have `cd ..` somewhere, suggesting that this "should occur" at the root folder

By using a separate `package.json` we can:

- Install firebase-related libraries as explicit dependencies and version control them (instead of a global install)
- Keep many deployment-related scripts in one place, instead of documenting and writing the whole commands everytime. 
- Don't bloat the `functions` `package.json` with dependencies and scripts that are not related to the actual functions code.

Why do you say that we need it inside the functions folder?

File: functions/src/shared/schemas/updateDiscussion.ts (hunk 2 above)
Comment (pepoospina) [reply functions/src/shared/schemas/updateDiscussion.ts:5]: Will change

File: functions/src/shared/schemas/hideContent.ts (hunk 3 above)
Comment (pepoospina) [reply functions/src/shared/schemas/hideContent.ts:6]: You can get the `commonId` from the `itemId` and `type`. Leaving it is risky as it may open attack surfaces, where a user sends a manipulated `commonId`. 

This also applies to the `userId`. I would remove it from all payloads since it is overwritten by `req.user.uid` everywhere.
